ambry-store/src/main/java/com.github.ambry.store/DiskSpaceAllocator.java

```diff
@@ -0,0 +1,419 @@
+ * {@link BlobStore}s and {@link BlobStoreCompactor}s. The disk segments will be preallocated on platforms that support
+ * the fallocate syscall. The entities that require disk space register there requirements using the
+ * {@link #initializePool(Collection)} method. Disk space is allocated using the
+ * {@link DiskSpaceAllocator#allocate(File, long)} method. If a segment matching the requested size is not in the pool,
+ * it will be created at runtime. Disk space is returned to the pool using the
+ * {@link DiskSpaceAllocator#free(File, long)} method. It is up to the requester to keep track of the size of the
+ * segment requested.
+ *
+ * NOTE: Segments can be allocated and freed before pool initialization.  This will fall back to preallocating at
+ * runtime if the current pool does not include a segment of the correct size.
+ */
+class DiskSpaceAllocator {
+  private static final Logger logger = LoggerFactory.getLogger(DiskSpaceAllocator.class);
+  private static final String RESERVE_FILE_PREFIX = "reserve_";
+  private static final String FILE_SIZE_DIR_PREFIX = "reserve_size_";
+  private static final FileFilter RESERVE_FILE_FILTER =
+      pathname -> pathname.isFile() && pathname.getName().startsWith(RESERVE_FILE_PREFIX);
+  private static final FileFilter FILE_SIZE_DIR_FILTER =
+      pathname -> pathname.isDirectory() && getFileSizeForDirName(pathname.getName()) != null;
+  private final File reserveDir;
+  private final long requiredSwapSegmentsPerSize;
+  private final StorageManagerMetrics metrics;
+  private ReserveFileMap reserveFiles;
+  private PoolState poolState = PoolState.NOT_INVENTORIED;
+
+  /**
+   * This constructor will  inventory any existing reserve files in the pool (adding them to the in-memory map) but not
+   * create the reserve directory if it does not exist. Any files that already exist in the pool can be checked out
+   * prior to calling {@link #initializePool(Collection)}. If a file of the correct size does not yet exist, a new file
+   * will be allocated at runtime.
+   * @param reserveDir the directory where reserve files will reside. If this directory does not exist yet, it will
+   *                   be created.
+   * @param requiredSwapSegmentsPerSize the number of swap segments needed for each segment size in the pool.
+   * @param metrics a {@link StorageManagerMetrics} instance.
+   * @throws StoreException
+   */
+  DiskSpaceAllocator(File reserveDir, long requiredSwapSegmentsPerSize, StorageManagerMetrics metrics) {
+    this.reserveDir = reserveDir;
+    this.requiredSwapSegmentsPerSize = requiredSwapSegmentsPerSize;
+    this.metrics = metrics;
+    try {
+      if (reserveDir.isDirectory()) {
+        reserveFiles = inventoryExistingReserveFiles();
+        poolState = PoolState.INVENTORIED;
+      }
+    } catch (Exception e) {
+      logger.error("Could not inventory preexisting reserve directory", e);
```

Comment:
The thinking was that the constructors called in DiskManager always succeed, and this will be re-attempted in `initializePool` if it fails here. If it throws an exception there, we can mark the disk as down but not affect startup on other disks.